MethodSeparationFixer - Fix spacing around comments

SpacePossum opened the pull request:
RFC, at see: https://github.com/FriendsOfPHP/PHP-CS-Fixer/issues/1421 after https://github.com/symfony/symfony/pull/15890

Targeting master because the fixer does not exists in the 1.x branches.

keradus commented:
@GrahamCampbell you have rised an issue, please, could you check this PR?

GrahamCampbell commented on Symfony/CS/Tests/Fixer/Symfony/MethodSeparationFixerTest.php:
```diff
@@ -204,7 +323,6 @@ public function method444f1()
     }
 
     /**/
-
```
surely the test should pass with this line still there

GrahamCampbell replied:
if it doesn't, then the implementation is wrong imo

SpacePossum replied:
input
```php

    /**/
    public function method444g()
    {
    }
```

output
```php
    /**/
    public function method444g()
    {
    }
```

I don't see the problem, you want the blank line back? Then this PR is not needed in the first place ^-^

GrahamCampbell replied:
i'm saying shouldn't delete it if it exists

GrahamCampbell replied:
that was the whole issue

SpacePossum replied:
Yes, but that is not what is tested there.
The test was `make sure a blank line is added if there is a comment (not a PHPDoc) above the method`, this is now no longer wanted right? So I've updated the test.

SpacePossum replied:
I've added a test, hope it is what you want:
https://github.com/SpacePossum/PHP-CS-Fixer/commit/7d849f39e0e60ec3b5612869bd30c484848b6c66

(reverted the other changes in the next commit)

GrahamCampbell replied:
:-1:

GrahamCampbell replied:
This still hasn't been reverted yet?

GrahamCampbell replied:
Both with and without should pass surely?

SpacePossum replied:
I've added a test for you case, so I'm not sure what you want me to add, remove and/or revert here?

GrahamCampbell replied:
This case essentially needs to stay unmodified and still pass.

gharlan replied:
@GrahamCampbell this is the "expected" part, input is here: https://github.com/SpacePossum/PHP-CS-Fixer/blob/master_1421_method_separation_tweaks/Symfony/CS/Tests/Fixer/Symfony/MethodSeparationFixerTest.php#L382-L385

So you want that the newline is added?

GrahamCampbell replied:
> @GrahamCampbell this is the "expected" part

Exactly. The expectation hasn't changed for the output. This fixer is broken if it removes that space.

gharlan replied:
> This fixer is broken if it removes that space.

Which space is removed? Again: In input there is no newline after the comment. And the fixer should not add a newline. This is fixed by this pr, so the test has to be adjusted.

SpacePossum replied:
More tests added;
https://github.com/SpacePossum/PHP-CS-Fixer/commit/e4f96e19d41879b3045c0e0357a267246f3ab81d

I really cannot think of other cases to add, so please provide those if want. Please do so in PHP code.

SpacePossum replied:
ping @GrahamCampbell , please have a fresh look at the tests. I would like to wrap this one up and backport it to the 1.x line.

SpacePossum replied:
>i'm saying shouldn't delete it if it exists

expected:
https://github.com/FriendsOfPHP/PHP-CS-Fixer/pull/1426/files#diff-57e53faa7c8496e4e9ff143e74804f78R54

input:
https://github.com/FriendsOfPHP/PHP-CS-Fixer/pull/1426/files#diff-57e53faa7c8496e4e9ff143e74804f78R90

this one is ready to go IMO

keradus commented:
what is the state of this PR?

SpacePossum commented:
Wait for feedback. I don't like to tag people, but maybe it would be useful to get the reporters in here as well (Stof should be back next week I think)

keradus commented:
@GrahamCampbell I remember that you play with the fixer on Symfony repo, could you discuss with @SpacePossum and then both of you could get this PR done smooth and shine

SpacePossum commented:
Rebased, anyone who would like to take a look?

SpacePossum commented:
rebased an squashed

gharlan commented on Symfony/CS/Fixer/Symfony/MethodSeparationFixer.php:
```diff
@@ -156,6 +156,38 @@ private function fixSpaceAboveMethod(Tokens $tokens, $classStart, $methodIndex)
+                        $this->correctLineBreaks($tokens, $nonWhiteAbove, $firstMethodAttrIndex, 2);
+                    }
+                }
+            }
```
The `if{if-else{if-else{if-else}}}`-construct is hard to read ;)

I would prefer some more `return`s instead of the `else`s.

SpacePossum replied:
:+1: 
https://github.com/SpacePossum/PHP-CS-Fixer/commit/172edb77019e79637d67a5081e14f342131ee2a1 let me know if I misinterpreted your suggestion

gharlan replied:
:+1: thanks!

gharlan commented:
:+1:

keradus commented:
One more :+1: and good to merge!

GrahamCampbell commented on Symfony/CS/Fixer/Symfony/MethodSeparationFixer.php:
```diff
@@ -156,6 +159,42 @@ private function fixSpaceAboveMethod(Tokens $tokens, $classStart, $methodIndex)
+            }
+
+            // $tokens[$nonWhiteAbove+1] is always a white space token here
+            $lineBreakCount = substr_count($tokens[$nonWhiteAbove + 1]->getContent(), "\n");
```
uneeded assignment

SpacePossum replied:
fixed

GrahamCampbell commented on Symfony/CS/Fixer/Symfony/MethodSeparationFixer.php:
```diff
@@ -156,6 +159,42 @@ private function fixSpaceAboveMethod(Tokens $tokens, $classStart, $methodIndex)
+                // no white space found between comment and method start
+                $this->correctLineBreaks($tokens, $nonWhiteAbove, $firstMethodAttrIndex, 1);
+
+                return;
```
can avoid a return here by just having elseif next time

keradus replied:
I would keep here the return.
@GrahamCampbell as you said few days ago - it's a matter of code style, and if we you almost everywhere early-return pattern, then your pointing to change them to if-elseif in every new PR is pointless. Both are good, but be consistence with codebase.

GrahamCampbell commented on Symfony/CS/Fixer/Symfony/MethodSeparationFixer.php:
```diff
@@ -156,6 +159,42 @@ private function fixSpaceAboveMethod(Tokens $tokens, $classStart, $methodIndex)
+                // more than one line break, always bring it back to 2 line breaks between the method start and what is above it
+                $this->correctLineBreaks($tokens, $nonWhiteAbove, $firstMethodAttrIndex, 2);
+
+                return;
```
same here

GrahamCampbell replied:
if
elseif
elseif
else

SpacePossum replied:
https://github.com/FriendsOfPHP/PHP-CS-Fixer/pull/1426#discussion_r45794736

>The if{if-else{if-else{if-else}}}-construct is hard to read ;)

https://github.com/FriendsOfPHP/PHP-CS-Fixer/pull/1497#discussion_r46095848
;)

anyhow, I'm always happy to refactor ^-^

gharlan replied:
I still prefer the `return`s. ;)
But the if-elseif-elseif-else would be better than the previous `if{if-else{if-else{if-else}}}`.
My main concern: avoid deep nesting.

keradus replied:
yep, if you have if-elseif-else then sth is terribly wrong...

GrahamCampbell replied:
it's just inconsistant, because you have an if if if else now

GrahamCampbell replied:
No need for that else.

SpacePossum replied:
> if you have if-elseif-else then sth is terribly wrong...

with with although respect, that sounds like a simplification and qualification/judgement that needs proof or holds not much value IMO.

GrahamCampbell replied:
Some people prefer if elseif elseif elseif elseif else over if return if return if return if return return.

I guess what you're saying is that you prefer the 2nd option, that's all.

GrahamCampbell replied:
I'd normally prefer the 1st option unless it casued any nesting, then i'd always go for an early break out. In this case, using elseif doesn't change the nesting.

SpacePossum replied:
>I guess what you're saying is that you prefer the 2nd option, that's all.

Not really, because I don't care much about it and I'm happy to refactor to whatever is needed.
I'm trying to say that while discussing what is a preference, I think we can do without qualifications as 'terribly wrong', since without arguments it doesn't add anything useful IMO.

I hope someone who knows what the code should like is willing to take over this PR, because I'm lost here/

GrahamCampbell commented on Symfony/CS/Fixer/Symfony/MethodSeparationFixer.php:
```diff
@@ -156,6 +159,42 @@ private function fixSpaceAboveMethod(Tokens $tokens, $classStart, $methodIndex)
+                $this->correctLineBreaks($tokens, $nonWhiteAboveComment, $nonWhiteAbove, $nonWhiteAboveComment === $classStart ? 1 : 2);
+            } else {
+                // 2. the comment belongs to the code above the method
+                //      make sure there is a blank line above the method (i.e. 2 line breaks)
```
funny indentation

SpacePossum commented:
Closing as it does not seem to be possible to find a consensuses about this one.

Kindly note @keradus that you might want to move the fixer to contrib (i.e. out of symfony level) as @ symfony they don't agree about the fixer behaviour.

keradus commented:
Reopen, will have a deeper look...

SpacePossum commented:
rebased
ping @keradus 
I would really like to get this one over with after 4 months of rebasing

keradus commented:
I found one last bug and good to go: https://github.com/SpacePossum/PHP-CS-Fixer/pull/8

keradus commented on src/Fixer/Symfony/MethodSeparationFixer.php:
```diff
@@ -156,6 +159,41 @@ private function fixSpaceAboveMethod(Tokens $tokens, $classStart, $methodIndex)
 
+        // deal with comments above a method
+        if ($tokens[$nonWhiteAbove]->isGivenKind(T_COMMENT)) {
+            if ($firstMethodAttrIndex - $nonWhiteAbove === 1) {
```
yoda

GrahamCampbell replied:
![image](https://cloud.githubusercontent.com/assets/2829600/12462634/4b1e5218-bfb7-11e5-9c86-725ec8925004.png)

SpacePossum replied:
will fix, btw. what happened to the yoda-fixer?

keradus replied:
is messed up and need to be rewritten

SpacePossum commented:
Thanks for checking @keradus and the bug report + test :+1: 
reworked.
There are still a few hits on symfony 2.3;
```
	modified:   src/Symfony/Bridge/Twig/Tests/Node/TransNodeTest.php
	modified:   src/Symfony/Component/Console/Tests/Descriptor/AbstractDescriptorTest.php
	modified:   src/Symfony/Component/CssSelector/Tests/Node/AbstractNodeTest.php
	modified:   src/Symfony/Component/CssSelector/Tests/Parser/Handler/AbstractHandlerTest.php
	modified:   src/Symfony/Component/DependencyInjection/Tests/Fixtures/php/services10.php
	modified:   src/Symfony/Component/DependencyInjection/Tests/Fixtures/php/services12.php
	modified:   src/Symfony/Component/DependencyInjection/Tests/Fixtures/php/services9_compiled.php
	modified:   src/Symfony/Component/Form/Tests/Extension/HttpFoundation/HttpFoundationRequestHandlerTest.php
	modified:   src/Symfony/Component/HttpFoundation/Response.php
	modified:   src/Symfony/Component/HttpKernel/Tests/Fragment/RoutableFragmentRendererTest.php
	modified:   src/Symfony/Component/Security/Tests/Core/Encoder/EncoderFactoryTest.php
	modified:   src/Symfony/Component/Validator/Tests/Mapping/Loader/AnnotationLoaderTest.php
```

I'll see if I PR those or maybe move the fixer to contrib level.

keradus commented:
for me it shows only 9 files, and if they are not autogenerated code they should be merged

keradus commented:
ready to merge

keradus commented:
Thank you @SpacePossum.

SpacePossum commented:
thanks all for reviewing and testing :)

keradus commented:
That was long, but we finally did it.
Great work @SpacePossum

SpacePossum commented:
>That was long, but we finally did it.

yeah, great to see this in, first fixer I wrote is accepted :D

keradus commented:
Really? It was your first fixer? no way!
